x/ccv/provider/keeper/relay.go

```diff
@@ -43,58 +26,95 @@ func removeStringFromSlice(slice []string, x string) (newSlice []string, numRemo
 	return newSlice, len(slice) - len(newSlice)
 }
 
-func (k Keeper) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, data ccv.ValidatorSetChangePacketData, ack channeltypes.Acknowledgement) error {
-	chainID, ok := k.GetChannelToChain(ctx, packet.DestinationChannel)
-	if !ok {
-		return sdkerrors.Wrapf(ccv.ErrInvalidConsumerChain, "chain ID doesn't exist for channel ID: %s", packet.DestinationChannel)
+// OnRecvVSCMaturedPacket handles a VSCMatured packet
+func (k Keeper) OnRecvVSCMaturedPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	data ccv.VSCMaturedPacketData,
+) exported.Acknowledgement {
+	// check that the channel is established
+	chainID, found := k.GetChannelToChain(ctx, packet.DestinationChannel)
+	if !found {
+		// VSCMatured packet was sent on a channel different than any of the established CCV channels
+		return utils.OnRecvPacketOnUnknownChannel(ctx, k.scopedKeeper, k.channelKeeper, packet)
 	}
 
+	// iterate over the unbonding operations mapped to (chainID, data.ValsetUpdateId)
 	unbondingOps, _ := k.GetUnbondingOpsFromIndex(ctx, chainID, data.ValsetUpdateId)
-
+	var maturedIds []uint64
 	for _, unbondingOp := range unbondingOps {
 		// remove consumer chain ID from unbonding op record
 		unbondingOp.UnbondingConsumerChains, _ = removeStringFromSlice(unbondingOp.UnbondingConsumerChains, chainID)
 
 		// If unbonding op is completely unbonded from all relevant consumer chains
 		if len(unbondingOp.UnbondingConsumerChains) == 0 {
-			// Attempt to complete unbonding in staking module
-			err := k.stakingKeeper.UnbondingCanComplete(ctx, unbondingOp.Id)
-			if err != nil {
-				return err
-			}
+			// Store id of matured unbonding op for later completion of unbonding in staking module
+			maturedIds = append(maturedIds, unbondingOp.Id)
 			// Delete unbonding op
 			k.DeleteUnbondingOp(ctx, unbondingOp.Id)
 		} else {
 			k.SetUnbondingOp(ctx, unbondingOp)
 		}
 	}
+	k.AppendMaturedUnbondingOps(ctx, maturedIds)
```

Comment:
Deferring completion to EndBlock is a divergence from the spec, we shouldn't forget that. 

- https://github.com/cosmos/ibc/issues/781